lib/semantics/tools.h

```diff
@@ -469,5 +469,43 @@ FindPolymorphicAllocatableUltimateComponent(const DerivedTypeSpec &);
 UltimateComponentIterator::const_iterator
 FindPolymorphicAllocatableNonCoarrayUltimateComponent(const DerivedTypeSpec &);
 
+class LabelEnforce {
+public:
+  LabelEnforce(SemanticsContext &context, std::set<parser::Label> &&labels,
+      parser::CharBlock constructSourcePosition, const char *construct)
+    : context_{context}, labels_{labels},
+      constructSourcePosition_{constructSourcePosition}, construct_{construct} {
+  }
+  template<typename T> bool Pre(const T &) { return true; }
+  template<typename T> bool Pre(const parser::Statement<T> &statement) {
+    currentStatementSourcePosition_ = statement.source;
+    return true;
+  }
+
+  template<typename T> void Post(const T &) {}
+
+  void Post(const parser::GotoStmt &gotoStmt);
+  void Post(const parser::ComputedGotoStmt &computedGotoStmt);
+  void Post(const parser::ArithmeticIfStmt &arithmeticIfStmt);
+  void Post(const parser::AssignStmt &assignStmt);
+  void Post(const parser::AssignedGotoStmt &assignedGotoStmt);
+  void Post(const parser::AltReturnSpec &altReturnSpec);
+  void Post(const parser::ErrLabel &errLabel);
+  void Post(const parser::EndLabel &endLabel);
+  void Post(const parser::EorLabel &eorLabel);
+  void checkLabelUse(const parser::Label &labelUsed);
+
+private:
+  SemanticsContext &context_;
+  std::set<parser::Label> labels_;
+  parser::CharBlock currentStatementSourcePosition_{nullptr};
+  parser::CharBlock constructSourcePosition_{nullptr};
+  const char *construct_{nullptr};
+
+  parser::MessageFormattedText GetEnclosingConstructMsg();
+  void SayWithConstruct(SemanticsContext &context,
+      parser::CharBlock stmtLocation, parser::MessageFormattedText &&message,
+      parser::CharBlock constructLocation);
+};  // class LabelEnforce
```

Comment:
We typically don't comment closing braces of classes or namespaces or functions.

Reply:
OK, I will remove this.